### PR TITLE
Service/AM: convert to ServiceFramework

### PR DESCRIPTION
**Pull request by wwylele**
@shinyquagsire23 request for review. Thanks!

Also removed/hide some deprecated ipc helper functions. The ones that are not removed are used by frd and nim.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3462)

### CODE REVIEW

File: src/core/hle/service/am/am.cpp [hunk 1]
```diff
@@ -590,58 +583,61 @@ void ListDLCContentInfos(Service::Interface* self) {
...
 
-void GetProgramList(Service::Interface* self) {
-    IPC::RequestParser rp(Kernel::GetCommandBuffer(), 2, 2, 2); // 0x00020082
+void Module::Interface::GetProgramList(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 2, 2, 2); // 0x00020082
 
     u32 count = rp.Pop<u32>();
     u8 media_type = rp.Pop<u8>();
-    VAddr title_ids_output_pointer = rp.PopMappedBuffer(nullptr);
+    auto& title_ids_output_pointer = rp.PopMappedBuffer();
```
Comment (B3n30) [inline src/core/hle/service/am/am.cpp]: title_ids_output

File: src/core/hle/service/am/am.cpp [hunk 2]
```diff
@@ -590,58 +583,61 @@ void ListDLCContentInfos(Service::Interface* self) {
...
-                media_type, title_id, content_count, content_ids_in);
+    rb.PushMappedBuffer(content_ids_in);
+    LOG_WARNING(Service_AM, "(STUBBED) media_type=%u, title_id=0x%016" PRIx64 ", content_count=%u",
+                media_type, title_id, content_count);
 }
 
-void GetProgramList(Service::Interface* self) {
-    IPC::RequestParser rp(Kernel::GetCommandBuffer(), 2, 2, 2); // 0x00020082
+void Module::Interface::GetProgramList(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 2, 2, 2); // 0x00020082
```
Comment (B3n30) [inline src/core/hle/service/am/am.cpp]: Nit: rp(ctx, **0x2**, 2, 2); same for some other functions

File: src/core/hle/service/am/am.cpp [hunk 3]
```diff
@@ -659,37 +655,34 @@ ResultCode GetTitleInfoFromList(const std::vector<u64>& title_id_list,
...
-    rb.PushMappedBuffer(title_id_list_pointer, title_id_list_size, title_id_list_perms);
-    rb.PushMappedBuffer(title_info_out, title_info_size, title_info_perms);
+    rb.PushMappedBuffer(title_id_list_buffer);
+    rb.PushMappedBuffer(title_info_out);
 }
 
-void DeleteUserProgram(Service::Interface* self) {
-    IPC::RequestParser rp(Kernel::GetCommandBuffer(), 0x000400, 3, 0);
+void Module::Interface::DeleteUserProgram(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x000400, 3, 0);
```
Comment (B3n30) [inline src/core/hle/service/am/am.cpp]: shouldn't that be 0x4?

File: src/core/hle/ipc_helpers.h [hunk 4]
```diff
@@ -311,8 +293,10 @@ class RequestParser : public RequestHelperBase {
 
     /**
      * Pop a descriptor containing `N` handles and resolves them to Kernel::Object pointers. If a
-     * handle is invalid, null is returned for that object instead. The same caveats from
-     * PopHandles() apply regarding `N` matching the number of handles in the descriptor.
+     * handle is invalid, null is returned for that object instead. The descriptor must contain
+     * exactly `N` handles, it is not permitted to, for example, call PopGenericObjects<1>() twice
+     * to read a multi-handle descriptor with 2 handles, or to make a single GenericObjects<2>()
```
Comment (Subv) [inline src/core/hle/ipc_helpers.h]: PopGenericObjects

**Comment by wwylele** [comment]
> Why was PopHandle renamed to PopHLEHandles though?

Because the "handle" used in the HLE IPC protocol is no longer a kernel handle type; it is an index into a object list instead, in which sense it is a "HLE handle" type. I expected that the IPC helper will be used only on top of ServiceFramework, in which case it never involves real handles. I want to emphasizers this so that readers won't think that kernel handles are still passing in the IPC helper.

**Pull request merged**